Feat/#21_#57 search for plants - Testing

tobireith opened the pull request:
- Fix Example Test, add config for running all tests
- Adjust Search to testing. Implement basic test for search functionality
- Remove unused stuff and TODO

tobireith committed:
Fix Example Test, add config for running all tests

tobireith committed:
Adjust Search to testing. Implement basic test for search functionality

tobireith committed:
Remove unused stuff and TODO

emuguy1 reviewed (approved):
In general, you shouldn't import multiple View Models into One Activity. The way to do this would be, to add the database connection into your own View Model but other than that. Looks great to me.

hoffmatteo commented on app/src/main/java/de/othr/plantico/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,11 @@
+
+import androidx.lifecycle.ViewModel
+import de.othr.plantico.database.entities.Plant
+
```
This ViewModel does not make sense, a ViewModel should hold the data itself. If you look at PlantViewModel, the ViewModel is created by a ViewModelFactory which passes the database repository, and the ViewModel holds the plant lists. If the data is passed by the PlantViewModel anyway, I don't see the reason for another ViewModel.

tobireith replied:
Deleted the ViewModel and put the search function to the PlantViewModel as discussed.
Please have a look again :)

tobireith committed:
Merge branch 'develop' into feat/#21-Search-for-Plants

tobireith committed:
Update Tests to JUnit4

tobireith committed:
Add JUnit4 TestRuntime

tobireith committed:
Make search method static, Revert 2nd view Model in search

By making the search method a static method it becomes easier to test.

tobireith committed:
Rename test file & class name

tobireith committed:
Fix search test

hoffmatteo commented on app/src/main/java/de/othr/plantico/ui/PlantViewModel.kt:
```diff
@@ -20,6 +20,13 @@ class TestViewModel(private val repository: Repository) : ViewModel() {
     fun insertOwnedPlant(plant: OwnedPlant)= viewModelScope.launch {
         repository.insertOwnedPlant(plant)
     }
+
+    companion object {
+        //Search for plants that contain the query as a substring. Not case-sensitive!
+        fun searchForPlantsInList(plants: List<Plant>, query: String): List<Plant> {
+            return plants.filter { plant: Plant -> plant.plantName.contains(query, ignoreCase = true) }
+        }
+    }
```
This does not make sense. A ViewModel contains the data, in this case the plantList, and passes it to the Activity. Here, you reverse this order, which is unnecessary as the ViewModel already has access to the plantList. Either use the ViewModel's data here, or move it to the Activity. In the Activity it shouldn't be static as well, as you should use the plantList property you defined there.

emuguy1 commented on app/build.gradle:
```diff
@@ -52,6 +52,7 @@ dependencies {
     implementation 'androidx.lifecycle:lifecycle-livedata-ktx:2.4.1'
     implementation 'androidx.lifecycle:lifecycle-viewmodel-ktx:2.4.1'
     testImplementation 'junit:junit:4.13.2'
+    testRuntimeOnly 'org.junit.vintage:junit-vintage-engine:5.8.2'
```
warum hast du da jUnit5 nochmal drinnen

emuguy1 reviewed (commented):
Agree to the comment of Matteo

tobireith committed:
Merge branch 'develop' into feat/#21-Search-for-Plants

# Conflicts:
#	app/build.gradle
#	app/src/main/java/de/othr/plantico/ui/PlantViewModel.kt
#	app/src/main/java/de/othr/plantico/ui/search/SearchableActivity.kt
#	app/src/test/java/de/othr/plantico/ExampleUnitTest.kt

tobireith committed:
Revert Companion Object for Plant Search

tobireith committed:
Adjust Test for SearchableActivity.kt

The test is now configured to test the searchForPlantsInList Method of the Searchable Activity.
